cmd/clusterctl/client/cluster/internal/dryrun/client.go

```diff
@@ -0,0 +1,354 @@
+	if uList, ok := list.(*unstructured.UnstructuredList); ok {
+		gvk = uList.GroupVersionKind()
+	} else {
+		var err error
+		gvk, err = apiutil.GVKForObject(list, c.fakeClient.Scheme())
+		if err != nil {
+			return errors.Wrap(err, "failed to get GVK of target object")
+		}
+	}
+
+	// Fetch lists from both fake client and the apiReader and merge the two lists.
+	uFakeList := &unstructured.UnstructuredList{}
+	uReaderList := &unstructured.UnstructuredList{}
+	uFakeList.SetGroupVersionKind(gvk)
+	uReaderList.SetGroupVersionKind(gvk)
+	if err := c.fakeClient.List(ctx, uFakeList, opts...); err != nil {
+		return err
+	}
+	if c.apiReader != nil {
+		if err := c.apiReader.List(ctx, uReaderList, opts...); err != nil {
+			return err
+		}
+		mergeLists(uFakeList, uReaderList)
+	}
+
+	if err := c.Scheme().Convert(uFakeList, list, nil); err != nil {
+		return errors.Wrapf(err, "failed to convert unstructured list to %T", list)
+	}
+	return nil
+}
+
+// Create saves the object in the internal object tracker.
+func (c *Client) Create(ctx context.Context, obj client.Object, opts ...client.CreateOption) error {
+	err := c.fakeClient.Create(ctx, obj, opts...)
+	if err == nil {
+		c.changeTracker.created[trackerIDOf(obj)] = true
+	}
+	return err
+}
+
+// Delete deletes the given obj from internal object tracker.
+// Delete will not affect objects in the Kubernetes Cluster.
+func (c *Client) Delete(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error {
+	err := c.fakeClient.Delete(ctx, obj, opts...)
+	if err == nil {
+		c.changeTracker.deleted[trackerIDOf(obj)] = true
+	} else if apierrors.IsNotFound(err) {
+		if c.apiReader == nil {
+			return err
+		}
+		// It is possible that we are trying to delete an object that exists in the Kubernetes Cluster but
+		// not in the internal object tracker.
+		// In such cases, check if the underlying object exists and track the operation. If the object does
+		// not exists return the original error.
+		tmpObj := obj.DeepCopyObject().(client.Object)
+		if getErr := c.apiReader.Get(ctx, client.ObjectKeyFromObject(obj), tmpObj); getErr != nil {
+			if apierrors.IsNotFound(getErr) {
+				// Deleted was called on an object that does no exists in the internal object tracker and in the
+				// Kubernetes Cluster. Return error.
+				// Note: return the original delete error. Not the get error.
+				return err
+			}
+			return errors.Wrap(err, "failed to check if object exists in underlying cluster")
+		}
+		// The objects exists in the Kubernetes Cluster. Track the delete operation.
+		c.changeTracker.deleted[trackerIDOf(obj)] = true
+		return nil
+	}
+	return err
+}
+
+// Update updates the given obj in the internal object tracker.
+// NOTE: Topology reconciler does not use update, so we are skipping implementation for now.
+func (c *Client) Update(ctx context.Context, obj client.Object, opts ...client.UpdateOption) error {
+	panic("Update method is not supported by the dryrun client")
```

Comment:
It has to be a panic. If the client does not support the function and the reconciler is using it we have to panic here. It we return an error the reconciler could have error handling and it would result in unexpected behavior.